pkg/kubelet/cm/memorymanager/state/state_mem.go

```diff
@@ -0,0 +1,127 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package state
+
+import (
+	"sync"
+
+	"k8s.io/klog"
+)
+
+type stateMemory struct {
+	sync.RWMutex
+	assignments  ContainerMemoryAssignments
+	machineState MemoryMap
+}
+
+var _ State = &stateMemory{}
+
+// NewMemoryState creates new State for keeping track of cpu/pod assignment
+func NewMemoryState() State {
+	klog.Infof("[memorymanager] initializing new in-memory state store")
+	return &stateMemory{
+		assignments:  ContainerMemoryAssignments{},
+		machineState: MemoryMap{},
+	}
+}
+
+// GetMemoryState returns Memory Map that stored in State
+func (s *stateMemory) GetMachineState() MemoryMap {
+	s.RLock()
+	defer s.RUnlock()
+
+	return s.machineState.Clone()
+}
+
+// GetMemoryBlocks returns memory assignments of container
+func (s *stateMemory) GetMemoryBlocks(podUID string, containerName string) []Block {
+	s.RLock()
+	defer s.RUnlock()
+
+	if res, ok := s.assignments[podUID][containerName]; ok {
+		return append([]Block{}, res...)
+	}
+	return nil
+}
+
+// GetMemoryAssignments returns ContainerMemoryAssignments
+func (s *stateMemory) GetMemoryAssignments() ContainerMemoryAssignments {
+	s.RLock()
+	defer s.RUnlock()
+
+	return s.assignments.Clone()
+}
+
+// SetMachineState stores MemoryMap in State
+func (s *stateMemory) SetMachineState(memoryMap MemoryMap) {
+	s.Lock()
+	defer s.Unlock()
+
+	s.machineState = memoryMap
```

Comment:
I was wondering whether we also need to clone `memoryMap` here... What do you think?
Or is there any unspoken assumption that an object resulting from `GetMachineState()` is to be immediately modified and promptly passed to `SetMachineState`? So that:
- the object is not modified by other threads (thread safe data sharing)
- and therefore there is no need to repeat cloning

Reply:
The common flow is:
- get machine state or assignments(you get clone)
- modify it
- set it

To be honest, I can not imagine a problematic situation(from threads perspective) here, can you add the problematic scenario that you see without additional clone?

Reply:
I didn't think about particular scenario, just a theoretical one in which a couple of threads operate on the same object (that was returned from `GetMachineState()`). But, it seems that in our case there will be no such a sharing, so: never mind :)